src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionHelper.cs

```diff
@@ -67,190 +68,187 @@ async Task<(ImmutableArray<SerializableImportCompletionItem>, StatisticCounter)>
-            using var _2 = PooledDictionary<INamespaceSymbol, string>.GetInstance(out var namespaceNameCache);
-
             // This dictionary is used as cache among all projects and PE references. 
             // The key is the receiver type as in the extension method declaration (symbol retrived from current compilation).
             // The value indicates if we can reduce an extension method with this receiver type given receiver type.
-            using var _3 = PooledDictionary<ITypeSymbol, bool>.GetInstance(out var checkedReceiverTypes);
-
+            var checkedReceiverTypes = new ConcurrentDictionary<ITypeSymbol, bool>();
+            var receiverTypeNames = GetReceiverTypeNames(receiverTypeSymbol);
             receiverTypeNames = AttachComplexTypes(receiverTypeNames);
 
-            // Get extension method items from source
-            foreach (var (project, syntaxIndex) in indices.SyntaxIndices!)
-            {
-                var filter = CreateAggregatedFilter(receiverTypeNames, syntaxIndex);
-                var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
-                var assembly = compilation.Assembly;
+            using var _ = ArrayBuilder<Task<ImmutableArray<IMethodSymbol>?>>.GetInstance(out var tasks);
 
-                var internalsVisible = currentAssembly.IsSameAssemblyOrHasFriendAccessTo(assembly);
-
-                var matchingMethodSymbols = GetPotentialMatchingSymbolsFromAssembly(
-                    compilation.Assembly, filter, namespaceFilter, internalsVisible,
-                    counter, cancellationToken);
-
-                if (project == currentProject)
-                {
-                    GetExtensionMethodItemsForSymbolsFromSameCompilation(
-                        position, semanticModel, receiverTypeSymbol, matchingMethodSymbols,
-                        completionItemsbuilder, namespaceNameCache, checkedReceiverTypes, cancellationToken);
-                }
-                else
-                {
-                    GetExtensionMethodItemsForSymbolsFromDifferentCompilation(
-                        position, semanticModel, receiverTypeSymbol, matchingMethodSymbols,
-                        completionItemsbuilder, namespaceNameCache, checkedReceiverTypes, cancellationToken);
-                }
+            foreach (var peReference in GetAllRelevantPeReferences(currentProject))
+            {
+                tasks.Add(Task.Run(() => GetExtensionMethodSymbolsFromPeReferenceAsync(
+                    peReference,
+                    forceIndexCreation,
+                    solution,
+                    semanticModel,
+                    receiverTypeSymbol,
+                    receiverTypeNames,
+                    position,
+                    namespaceInScope,
+                    checkedReceiverTypes,
+                    cancellationToken), cancellationToken));
```

Comment:
for the future.  when you're at this many passed parameters.  it's time to make a helper class that you isntantiate and move logic to.